### PR TITLE
Adding support for QConv2DBatchnorm

### PR DESCRIPTION
**Pull request by nicologhielmetti**
Here a first version for supporting QConv2DBatchnorm layer in hls4ml.

### CODE REVIEW

File: hls4ml/converters/keras/qkeras_layers.py [hunk 1]
```diff
@@ -100,3 +100,9 @@ def parse_qbatchnorm_layer(keras_layer, input_names, input_shapes, data_reader,
 
     return layer, output_shape
 
+
+@keras_handler('QConv2DBatchnorm')
+def parse_qconv2dbatchnorm_layer(keras_layer, input_names, input_shapes, data_reader, config):
+    conv_layer, intermediate_shape = parse_qconv_layer(keras_layer, input_names, input_shapes, data_reader, config)
+    batch_layer, out_shape = parse_batchnorm_layer(keras_layer, input_names, intermediate_shape, data_reader, config)
+    return {**conv_layer, **batch_layer}, out_shape
```
Comment (thesps) [inline hls4ml/converters/keras/qkeras_layers.py:108]: Looks good

File: hls4ml/converters/keras/qkeras_layers.py [hunk 2]
```diff
@@ -90,7 +90,7 @@ def parse_qactivation_layer(keras_layer, input_names, input_shapes, data_reader,
 
 @keras_handler('QBatchNormalization')
 def parse_qbatchnorm_layer(keras_layer, input_names, input_shapes, data_reader, config):
-    
+
```
Comment (thesps) [inline hls4ml/converters/keras/qkeras_layers.py:93]: Undo these formatting changes

File: hls4ml/converters/keras/qkeras_layers.py [hunk 3]
```diff
@@ -28,26 +28,26 @@ def parse_qdense_layer(keras_layer, input_names, input_shapes, data_reader, conf
 @keras_handler('QConv1D', 'QConv2D')
 def parse_qconv_layer(keras_layer, input_names, input_shapes, data_reader, config):
     assert('QConv' in keras_layer['class_name'])
-    
-    if int(keras_layer['class_name'][-2]) == 1:
+
+    if "1D" in keras_layer['class_name']:
         layer, output_shape = parse_conv1d_layer(keras_layer, input_names, input_shapes, data_reader, config)
-    elif int(keras_layer['class_name'][-2]) == 2:
+    elif "2D" in keras_layer['class_name']:
```
Comment (thesps) [inline hls4ml/converters/keras/qkeras_layers.py:34]: These changes are not specific to the QConv2DBatchnorm - please remove them, but consider making a new PR with these

Comment (thesps) [inline hls4ml/converters/keras/qkeras_layers.py:34]: Is this because the '2' `QConv2DBatchnorm` is not at position `[-2]`?

File: hls4ml/converters/keras/convolution.py [hunk 4]
```diff
@@ -84,5 +84,8 @@ def parse_conv2d_layer(keras_layer, input_names, input_shapes, data_reader, conf
         output_shape = [input_shapes[0][0], layer['n_filt'], layer['out_height'], layer['out_width']]
     else:
         output_shape = [input_shapes[0][0], layer['out_height'], layer['out_width'], layer['n_filt']]
-
+    if layer['class_name'] == 'QConv2DBatchnorm':
+        intermediate_shape = list()
+        intermediate_shape.append(output_shape)
+        output_shape = intermediate_shape
```
Comment (thesps) [inline hls4ml/converters/keras/convolution.py:90]: Why is this needed?

File: hls4ml/converters/keras/core.py [hunk 5]
```diff
@@ -103,7 +103,7 @@ def parse_activation_layer(keras_layer, input_names, input_shapes, data_reader,
 
 @keras_handler('BatchNormalization')
 def parse_batchnorm_layer(keras_layer, input_names, input_shapes, data_reader, config):
-    assert('BatchNormalization' in keras_layer['class_name'])
+    assert('batchnorm' in keras_layer['class_name'].lower())
```
Comment (thesps) [inline hls4ml/converters/keras/core.py:106]: I suggest keeping the original condition, and adding the new condition in '`or`'

File: hls4ml/model/hls_layers.py [hunk 6]
```diff
@@ -6,6 +6,9 @@
 import numpy as np
 from collections import OrderedDict
 
+from tensorflow.python.ops import math_ops
```
Comment (thesps) [inline hls4ml/model/hls_layers.py:9]: We'd like to avoid adding a new dependency. Can we use numpy?

File: hls4ml/model/hls_layers.py [hunk 7]
```diff
@@ -850,6 +855,112 @@ def function_cpp(self):
...
+        mult_params['n_in'] = self.get_attr('n_chan') * self.get_attr('filt_height') * self.get_attr('filt_width')
+        mult_params['n_out'] = self.get_attr('n_filt')
+        mult_params['product_type'] = self.model.config.backend.product_type(self.get_input_variable().type.precision,
+                                                                             self.get_weights('weight').type.precision)
+        mult_config = self._config_template[1].format(**mult_params)
+
+        return mult_config + '\n' + conv_config
+
+
+class Conv2DBatchnorm(Layer):
```
Comment (thesps) [inline hls4ml/model/hls_layers.py:906]: Instead, inherit from existing `Conv2D`, and call `initialize`

File: hls4ml/model/hls_layers.py [hunk 8]
```diff
@@ -850,6 +855,112 @@ def function_cpp(self):
...
+        kernel = self.model.get_weights_data(self.name, 'kernel')
+        bias = self.model.get_weights_data(self.name, 'bias')
+
+        # get batchnorm weights and moving stats
+        gamma = self.model.get_weights_data(self.name, 'gamma')
+        beta = self.model.get_weights_data(self.name, 'beta')
+        moving_mean = self.model.get_weights_data(self.name, 'moving_mean')
+        moving_variance = self.model.get_weights_data(self.name, 'moving_variance')
+        # get the inversion factor so that we replace division by multiplication
+        inv = math_ops.rsqrt(moving_variance + self.get_attr('epsilon'))
```
Comment (thesps) [inline hls4ml/model/hls_layers.py:924]: Can we replace this with something from numpy, for example?

File: hls4ml/model/hls_layers.py [hunk 9]
```diff
@@ -850,6 +855,112 @@ def function_cpp(self):
...
+        if gamma is not None:
+            inv *= gamma
+
+        # wrap conv kernel and bias with bn parameters
+        folded_kernel = inv * kernel
+        folded_bias = inv * (bias - moving_mean) + beta
+
+        return [folded_kernel, folded_bias]
+
+    def initialize(self):
```
Comment (thesps) [inline hls4ml/model/hls_layers.py:934]: After inheriting from `Conv2D` class, call its `initialize` function, then modify the weights inplace, like in existing line 951

File: hls4ml/model/hls_layers.py [hunk 10]
```diff
@@ -448,15 +451,17 @@ def make_stream_variable(self, shape, dim_names, var_name='layer{index}_out', ty
         
         return StreamVariable(shape, dim_names, var_name=var_name, type_name=type_name, precision=precision, n_pack=pack_factor, depth=depth, index=self.index)
 
-    def add_weights(self, quantizer=None, compression=False):
-        data = self.model.get_weights_data(self.name, 'kernel')
+    def add_weights(self, quantizer=None, compression=False, data=None):
```
Comment (thesps) [inline hls4ml/model/hls_layers.py:454]: After making below change `Conv2DBatchnorm(Layer)` -> `Conv2DBatchnorm(Conv2D)`, you won't need these any more.
`add_weights_variable` already has the functionality to add the data.

File: hls4ml/templates/vivado_template.py [hunk 11]
```diff
@@ -375,6 +407,7 @@ def __init__(self):
         self.register_templates('BatchNormalization'     , batchnorm_function_template,   batchnorm_config_template, batchnorm_include_list)
         self.register_templates('Conv1D'                 , conv1d_function_template,      [conv1d_config_template, conv_mult_config_template], conv1d_include_list)
         self.register_templates('Conv2D'                 , conv2d_function_template,      [conv2d_config_template, conv_mult_config_template], conv2d_include_list)
+        self.register_templates('Conv2DBatchnorm'        , conv2dbatchnorm_function_template, [conv2dbatchnorm_config_template, conv_mult_config_template], conv2dbatchnorm_include_list)
```
Comment (thesps) [inline hls4ml/templates/vivado_template.py:410]: Use the existing `Conv2D` templates, ie
`        self.register_templates('Conv2DBatchnorm'                 , conv2d_function_template,      [conv2d_config_template, conv_mult_config_template], conv2d_include_list)
`

File: hls4ml/templates/vivado_template.py [hunk 12]
```diff
@@ -357,6 +388,7 @@
 batchnorm_include_list = ['nnet_utils/nnet_batchnorm.h', 'nnet_utils/nnet_batchnorm_stream.h']
 conv1d_include_list = ['nnet_utils/nnet_conv1d.h', 'nnet_utils/nnet_conv1d_stream.h']
 conv2d_include_list = ['nnet_utils/nnet_conv2d.h', 'nnet_utils/nnet_conv2d_stream.h']
+conv2dbatchnorm_include_list = ['nnet_utils/nnet_conv2d.h', 'nnet_utils/nnet_conv2d_stream.h']
```
Comment (thesps) [inline hls4ml/templates/vivado_template.py:391]: Don't need to define new templates for this layer

File: hls4ml/templates/vivado_template.py [hunk 13]
```diff
@@ -336,6 +366,7 @@
 batchnorm_function_template = 'nnet::normalize<{input_t}, {output_t}, {config}>({input}, {output}, {scale}, {bias});'
 conv1d_function_template = 'nnet::conv_1d_{data_format}<{input_t}, {output_t}, {config}>({input}, {output}, {w}, {b});'
 conv2d_function_template = 'nnet::conv_2d_{data_format}<{input_t}, {output_t}, {config}>({input}, {output}, {w}, {b});'
+conv2dbatchnorm_function_template = 'nnet::conv_2d_{data_format}<{input_t}, {output_t}, {config}>({input}, {output}, {w}, {b});'
```
Comment (thesps) [inline hls4ml/templates/vivado_template.py:369]: Don't need to define new templates for this layer

File: hls4ml/templates/vivado_template.py [hunk 14]
```diff
@@ -101,6 +101,36 @@
...
+    static const unsigned min_height = {min_height};
+    static const unsigned min_width = {min_width};
+    static const ap_uint<filt_height * filt_width> pixels[min_height * min_width];
+    typedef {accum_t} accum_t;
+    typedef {bias_t} bias_t;
+    typedef {weight_t} weight_t;
+    typedef {config_t} mult_config;
+}};
+const ap_uint<config{index}::filt_height * config{index}::filt_width> config{index}::pixels[] = {{{instructions}}};\n"""
+
```
Comment (thesps) [inline hls4ml/templates/vivado_template.py:133]: Don't need to define new templates for this layer

File: hls4ml/converters/keras/convolution.py (hunk 4 above)
Comment (nicologhielmetti) [reply hls4ml/converters/keras/convolution.py:90]: Because for `QConv2DBatchnorm` `parse_conv2d_layer()` and `parse_batchnorm_layer()` are called one after the other. Without embedding the shape in a 2D-vector here, it will result in an error when `parse_batchnorm_layer()` executes, see line 111 of keras/core.py ( `for dim in input_shapes[0][1:]:` ). Otherwise I can add a control inside the `parse_batchnorm_layer()` method that checks if the considered layer is `QConv2DBatchnorm` and modify the way it access `input_shapes` field.

File: hls4ml/converters/keras/qkeras_layers.py (hunk 3 above)
Comment (nicologhielmetti) [reply hls4ml/converters/keras/qkeras_layers.py:34]: Yes, in this way it should be more general, accepting also `QConv2DBatchnorm`. Adding an '`or`' condition considering the case of `QConv2DBatchnorm` did not seem to me a good solution.

**Comment by nicologhielmetti** [comment]
I created a notebook jupyter as a [gist](https://gist.github.com/nicologhielmetti/6ce8af8e1a36693a6b587fbdf4440caf) for testing the integration.

Note: it requires `pip install git+https://github.com/nicologhielmetti/hls4ml.git@6e6afab0a106d5cecb1cf42a498ed5d7770aa252`
and `pip install git+https://github.com/google/qkeras.git@25e7803bcf2d1a4925f705c5678bd171a375f618` in addition to the packages in the following `conda env` file:

```
name: hls4ml-qconv2dbatchnorm
channels:
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - argon2-cffi=20.1.0=py37h27cfd23_1
  - async_generator=1.10=py37h28b3542_0
  - attrs=20.3.0=pyhd3eb1b0_0
  - backcall=0.2.0=pyhd3eb1b0_0
  - blas=1.0=mkl
  - bleach=3.3.0=pyhd3eb1b0_0
  - ca-certificates=2021.1.19=h06a4308_1
  - certifi=2020.12.5=py37h06a4308_0
  - cffi=1.14.5=py37h261ae71_0
  - dbus=1.13.18=hb2f20db_0
  - decorator=4.4.2=pyhd3eb1b0_0
  - defusedxml=0.6.0=pyhd3eb1b0_0
  - entrypoints=0.3=py37_0
  - expat=2.2.10=he6710b0_2
  - fontconfig=2.13.1=h6c09931_0
  - freetype=2.10.4=h5ab3b9f_0
  - glib=2.67.4=h36276a3_1
  - gst-plugins-base=1.14.0=h8213a91_2
  - gstreamer=1.14.0=h28cd5cc_2
  - h5py=2.10.0=py37hd6299e0_1
  - hdf5=1.10.6=hb1b8bf9_0
  - icu=58.2=he6710b0_3
  - importlib-metadata=2.0.0=py_1
  - importlib_metadata=2.0.0=1
  - intel-openmp=2020.2=254
  - ipykernel=5.3.4=py37h5ca1d4c_0
  - ipython=7.21.0=py37hb070fc8_0
  - ipython_genutils=0.2.0=pyhd3eb1b0_1
  - ipywidgets=7.6.3=pyhd3eb1b0_1
  - jedi=0.17.0=py37_0
  - jinja2=2.11.3=pyhd3eb1b0_0
  - jpeg=9b=h024ee3a_2
  - jsonschema=3.2.0=py_2
  - jupyter=1.0.0=py37_7
  - jupyter_client=6.1.7=py_0
  - jupyter_console=6.2.0=py_0
  - jupyter_core=4.7.1=py37h06a4308_0
  - jupyterlab_pygments=0.1.2=py_0
  - jupyterlab_widgets=1.0.0=pyhd3eb1b0_1
  - kiwisolver=1.3.1=py37h2531618_0
  - lcms2=2.11=h396b838_0
  - ld_impl_linux-64=2.33.1=h53a641e_7
  - libedit=3.1.20191231=h14c3975_1
  - libffi=3.3=he6710b0_2
  - libgcc-ng=9.1.0=hdf63c60_0
  - libgfortran-ng=7.3.0=hdf63c60_0
  - libpng=1.6.37=hbc83047_0
  - libsodium=1.0.18=h7b6447c_0
  - libstdcxx-ng=9.1.0=hdf63c60_0
  - libtiff=4.1.0=h2733197_1
  - libuuid=1.0.3=h1bed415_2
  - libxcb=1.14=h7b6447c_0
  - libxml2=2.9.10=hb55368b_3
  - lz4-c=1.9.3=h2531618_0
  - markupsafe=1.1.1=py37h14c3975_1
  - matplotlib=3.3.4=py37h06a4308_0
  - matplotlib-base=3.3.4=py37h62a2d02_0
  - mistune=0.8.4=py37h14c3975_1001
  - mkl=2020.2=256
  - mkl-service=2.3.0=py37he8ac12f_0
  - mkl_fft=1.3.0=py37h54f3939_0
  - mkl_random=1.1.1=py37h0573a6f_0
  - nbclient=0.5.3=pyhd3eb1b0_0
  - nbconvert=6.0.7=py37_0
  - nbformat=5.1.2=pyhd3eb1b0_1
  - ncurses=6.2=he6710b0_1
  - nest-asyncio=1.5.1=pyhd3eb1b0_0
  - notebook=6.2.0=py37h06a4308_0
  - numpy-base=1.19.2=py37hfa32c7d_0
  - olefile=0.46=py37_0
  - openssl=1.1.1j=h27cfd23_0
  - packaging=20.9=pyhd3eb1b0_0
  - pandoc=2.11=hb0f4dca_0
  - pandocfilters=1.4.3=py37h06a4308_1
  - parso=0.8.1=pyhd3eb1b0_0
  - pcre=8.44=he6710b0_0
  - pexpect=4.8.0=pyhd3eb1b0_3
  - pickleshare=0.7.5=pyhd3eb1b0_1003
  - pip=21.0.1=py37h06a4308_0
  - prometheus_client=0.9.0=pyhd3eb1b0_0
  - prompt-toolkit=3.0.8=py_0
  - prompt_toolkit=3.0.8=0
  - ptyprocess=0.7.0=pyhd3eb1b0_2
  - pycparser=2.20=py_2
  - pygments=2.8.0=pyhd3eb1b0_0
  - pyparsing=2.4.7=pyhd3eb1b0_0
  - pyqt=5.9.2=py37h05f1152_2
  - pyrsistent=0.17.3=py37h7b6447c_0
  - python=3.7.10=hdb3f193_0
  - python-dateutil=2.8.1=pyhd3eb1b0_0
  - pytz=2021.1=pyhd3eb1b0_0
  - pyyaml=5.4.1=py37h27cfd23_1
  - pyzmq=20.0.0=py37h2531618_1
  - qt=5.9.7=h5867ecd_1
  - qtconsole=5.0.2=pyhd3eb1b0_0
  - qtpy=1.9.0=py_0
  - readline=8.1=h27cfd23_0
  - scipy=1.6.1=py37h91f5cce_0
  - seaborn=0.11.1=pyhd3eb1b0_0
  - send2trash=1.5.0=pyhd3eb1b0_1
  - setuptools=52.0.0=py37h06a4308_0
  - sip=4.19.8=py37hf484d3e_0
  - six=1.15.0=py37h06a4308_0
  - sqlite=3.33.0=h62c20be_0
  - terminado=0.9.2=py37h06a4308_0
  - testpath=0.4.4=pyhd3eb1b0_0
  - tk=8.6.10=hbc83047_0
  - tornado=6.1=py37h27cfd23_0
  - traitlets=5.0.5=pyhd3eb1b0_0
  - wcwidth=0.2.5=py_0
  - webencodings=0.5.1=py37_1
  - wheel=0.36.2=pyhd3eb1b0_0
  - widgetsnbextension=3.5.1=py37_0
  - xz=5.2.5=h7b6447c_0
  - yaml=0.2.5=h7b6447c_0
  - zeromq=4.3.3=he6710b0_3
  - zipp=3.4.0=pyhd3eb1b0_0
  - zlib=1.2.11=h7b6447c_3
  - zstd=1.4.5=h9ceee32_0
  - pip:
    - absl-py==0.11.0
    - alabaster==0.7.12
    - astunparse==1.6.3
    - babel==2.9.0
    - cachetools==4.2.1
    - chardet==4.0.0
    - colorama==0.4.4
    - cycler==0.10.0
    - dm-tree==0.1.5
    - docutils==0.16
    - flatbuffers==1.12
    - future==0.18.2
    - gast==0.3.3
    - google-auth==1.27.0
    - google-auth-oauthlib==0.4.2
    - google-pasta==0.2.0
    - grpcio==1.32.0
    - hls4ml==0.5.0
    - idna==2.10
    - imagesize==1.2.0
    - joblib==1.0.1
    - keras-preprocessing==1.1.2
    - keras-tuner==1.0.2
    - markdown==3.3.4
    - networkx==2.5
    - numpy==1.18.5
    - oauthlib==3.1.0
    - onnx==1.8.1
    - opt-einsum==3.3.0
    - pandas==1.2.3
    - parse==1.6.5
    - pillow==8.1.1
    - protobuf==3.15.3
    - pyasn1==0.4.8
    - pyasn1-modules==0.2.8
    - pydot==1.4.2
    - pydotplus==2.0.2
    - pyparser==1.0
    - python-graphviz==0.16
    - requests==2.25.1
    - requests-oauthlib==1.3.0
    - rsa==4.7.2
    - scikit-learn==0.24.1
    - sklearn==0.0
    - snowballstemmer==2.1.0
    - sphinx==3.5.1
    - sphinx-rtd-theme==0.5.1
    - sphinxcontrib-applehelp==1.0.2
    - sphinxcontrib-devhelp==1.0.2
    - sphinxcontrib-htmlhelp==1.0.3
    - sphinxcontrib-jsmath==1.0.1
    - sphinxcontrib-qthelp==1.0.3
    - sphinxcontrib-serializinghtml==1.1.4
    - tabulate==0.8.9
    - tensorboard==2.4.1
    - tensorboard-plugin-wit==1.8.0
    - tensorflow==2.3.1
    - tensorflow-estimator==2.3.0
    - tensorflow-model-optimization==0.5.0
    - termcolor==1.1.0
    - terminaltables==3.1.0
    - threadpoolctl==2.1.0
    - tqdm==4.58.0
    - typing-extensions==3.7.4.3
    - urllib3==1.26.3
    - werkzeug==1.0.1
    - wrapt==1.12.1
```

File: hls4ml/converters/keras/convolution.py [hunk 15]
```diff
@@ -84,5 +84,4 @@ def parse_conv2d_layer(keras_layer, input_names, input_shapes, data_reader, conf
         output_shape = [input_shapes[0][0], layer['n_filt'], layer['out_height'], layer['out_width']]
     else:
         output_shape = [input_shapes[0][0], layer['out_height'], layer['out_width'], layer['n_filt']]
-
```
Comment (thesps) [inline hls4ml/converters/keras/convolution.py:87]: Can you revert this file to the original such that there is no extra newline?

File: hls4ml/converters/keras/core.py [hunk 16]
```diff
@@ -103,7 +103,7 @@ def parse_activation_layer(keras_layer, input_names, input_shapes, data_reader,
 
 @keras_handler('BatchNormalization')
 def parse_batchnorm_layer(keras_layer, input_names, input_shapes, data_reader, config):
-    assert('BatchNormalization' in keras_layer['class_name'])
+    assert('BatchNormalization' in keras_layer['class_name'] or 'QConv2DBatchnorm' in keras_layer['class_name'])
```
Comment (thesps) [inline hls4ml/converters/keras/core.py:106]: I think this is okay for the time being. In the end, it would be nice if the `BatchNormalization` layer converter does not need to know about the extra layer. For example, when we get `QDenseBatchnorm` as well it would need to go here. But the best solution for this would be if QKeras just renames the layers, so let's leave it like this for now and wait.

File: hls4ml/converters/keras_to_hls.py [hunk 17]
```diff
@@ -124,7 +124,7 @@ def parse_default_keras_layer(keras_layer, input_names):
     return layer
 
 def parse_data_format(input_shape, data_format='channels_last'):
-    if input_shape[0] is None:
+    if len(input_shape) == 4 or input_shape[0] is None:
```
Comment (thesps) [inline hls4ml/converters/keras_to_hls.py:127]: I think we can simply remove this check altogether. The first dimension is always the batch size, and should always be discarded. Right @vloncar ?

File: hls4ml/converters/keras/qkeras_layers.py [hunk 18]
```diff
@@ -29,9 +29,9 @@ def parse_qdense_layer(keras_layer, input_names, input_shapes, data_reader, conf
 def parse_qconv_layer(keras_layer, input_names, input_shapes, data_reader, config):
     assert('QConv' in keras_layer['class_name'])
     
-    if int(keras_layer['class_name'][-2]) == 1:
+    if "1D" in keras_layer['class_name']:
         layer, output_shape = parse_conv1d_layer(keras_layer, input_names, input_shapes, data_reader, config)
-    elif int(keras_layer['class_name'][-2]) == 2:
+    elif "2D" in keras_layer['class_name']:
```
Comment (thesps) [inline hls4ml/converters/keras/qkeras_layers.py:34]: On this, and two lines avove, we should use single quotes rather than double

File: hls4ml/converters/keras/convolution.py [hunk 19]
```diff
@@ -85,4 +85,4 @@ def parse_conv2d_layer(keras_layer, input_names, input_shapes, data_reader, conf
     else:
         output_shape = [input_shapes[0][0], layer['out_height'], layer['out_width'], layer['n_filt']]
 
-    return layer, output_shape
+    return layer, output_shape
```
Comment (thesps) [inline hls4ml/converters/keras/convolution.py:88]: Now there is a difference here. Again, it would be good if this file could be unchanged wrt master branch

File: hls4ml/model/hls_layers.py [hunk 20]
```diff
@@ -896,6 +896,58 @@ def config_cpp(self):
...
+        folded_kernel = inv * kernel
+        folded_bias = inv * (bias - moving_mean) + beta
+
+        return [folded_kernel, folded_bias]
+
+    def initialize(self):
+        super(Conv2DBatchnorm, self).initialize()
+        folded_weights, folded_bias = self._get_folded_weights()
+        if self.model.config.is_resource_strategy(self) and self.model.config.backend.name == 'Vivado':
+            self.weights['weight'].data_unquantized = np.transpose(folded_weights, axes=[3, 2, 0, 1])
```
Comment (vloncar) [inline hls4ml/model/hls_layers.py:934]: Why do we need to keep this around?

Comment (thesps) [reply hls4ml/model/hls_layers.py:934]: The QKerasFactorizeAlpha passes [uses it](https://github.com/fastmachinelearning/hls4ml/blob/master/hls4ml/model/optimizer/passes/qkeras.py#L131), for example

Comment (nicologhielmetti) [reply hls4ml/model/hls_layers.py:934]: I think that I should update the transpose axes to `axes=[3, 0, 1, 2]`

**Review by thesps — APPROVED** [review]
